[FastISel][AArch64] Compare Instruction Miscompilation Fix

brendaso1 opened the pull request:
When shl is folded in compare instruction, a miscompilation occurs when the CMP instruction is also sign-extended. For the following IR:

  %op3 = shl i8 %op2, 3
  %tmp3 = icmp eq i8 %tmp2, %op3

It used to generate

   cmp w8, w9, sxtb #3

which means sign extend w9, shift left by 3, and then compare with the value in w8. However, the original intention of the IR would require `%op2` to first shift left before extending the operands in the comparison operation . Moreover, if sign extension is used instead of zero extension, the sample test would miscompile. This PR creates a fix for the issue, more specifically to not fold the left shift into the CMP instruction, and to create a zero-extended value rather than a sign-extended value.

brendaso1 committed:
[FastISel][AArch64] compare instruction miscompilation fix

When shl is folded in compare instruction, a miscompilation occurs
when the CMP instruction is also sign-extended. This fix avoids
folding shl into cmp instruction when the second operand is
extended, and also turns equality operations into unsigned
comparisons.

davemgreen commented on llvm/test/CodeGen/AArch64/fastisel-shift-and-cmp.ll:
```diff
@@ -0,0 +1,33 @@
+@.str = private unnamed_addr constant [13 x i8] c"TRUE BRANCH\0A\00"
+@.str.1 = private unnamed_addr constant [14 x i8] c"FALSE BRANCH\0A\00"
+
+define dso_local i32 @main() {
```
Is it possible to simplify the test? I think it could
 - remove dso_local
 - remove the globals
 - remove the loads
 - remove the calls
 - remove --global-isel=false and -O0?

There are some existing tests in llvm/test/CodeGen/AArch64/arm64-fast-isel-icmp.ll, but they don't seem to include any with shift amounts at the moment.

brendaso1 replied:
hmm in that case do you think it's more beneficial if I put this test in llvm/test/CodeGen/AArch64/arm64-fast-isel-icmp.ll?

davemgreen replied:
Yeah that sounds like it might be a good place for it. Feel free to use the update_llc_test_checks script too to generate the check lines.

davemgreen commented on llvm/lib/Target/AArch64/AArch64FastISel.cpp:
```diff
@@ -1233,7 +1233,11 @@ unsigned AArch64FastISel::emitAddSub(bool UseAdd, MVT RetVT, const Value *LHS,
-        if ((SI->getOpcode() == Instruction::Shl) && (C->getZExtValue() < 4)) {
+        if ((SI->getOpcode() == Instruction::Shl) && (C->getZExtValue() < 4) &&
+            !NeedExtend) {
+          // We can only fold instructions that doesn't need extension because
```
Is it possible to reach this point now, if NeedExtend==false and ExtendType != AArch64_AM::InvalidShiftExtend?

brendaso1 replied:
I didn't add `ExtendType != AArch64_AM::InvalidShiftExtend` because there is already a check before this nested-if statement in line 1232. Do you think it's still better to add the check?

davemgreen replied:
Yeah - I meant that if `ExtendType != InvalidShift` then does that mean that NeedExtend is always false? Is it ever possible to get to the emitAddSub_rx(..) line now?

brendaso1 replied:
Ah Gotcha. You're right, if `ExtendType != AArch64_AM::InvalidShiftExtend`, NeedExtend is always true. So the logic would never be reached. Maybe it's a better idea to remove that part of the logic all together?

brendaso1 committed:
Simplify test/CodeGen/AArch64/fastisel-shift-and-cmp.ll

- remove dso_local
- remove the globals
- remove the loads
- remove the calls
- remove --global-isel=false and -O0

brendaso1 committed:
removed extraneous function in unit test

brendaso1 committed:
clean up unit test

brendaso1 commented:
I have a check that is failing but I am not sure what the cause is.... It looks like the last tests that failed are on MLIR: https://buildkite.com/llvm-project/github-pull-requests/builds/25125 , but my fix has nothing to do with that part of the code, and all the automated checks were passing a few days ago. Is it because this branch does not have the most up-to-date commits as main? Should I merge / rebase my branch to main in this case?

davemgreen commented:
Yep I think we can ignore those errors, they look unrelated.

brendaso1 committed:
Removed left shift and compare merging logic because that cannot be reached

davemgreen reviewed (approved):
Can you use the update_llc_test_checks.py script to update the test? Then this LGTM, thanks.

brendaso1 committed:
moved unit test to arm64-fast-isel-icmp.ll

brendaso1 committed:
removed unnecessary changes to selectBranch

brendaso1 commented:
Added the generated CHECK lines to the unit tests. Let's see how the builders look.

Also, when I did more testing, it seems like the sample test wouldn't fail even if the Cmp is sign-extended rather than zero-extended in assembly. So I removed the related changes as well.

davemgreen commented:
Thanks. This LGTM. Are you happy for it to be squashed and merged?

brendaso1 commented:
Yup I will squash and merge

aemerson commented:
Thanks for the fix @brendaso1. I find the "squash and merge" button satisfying so I hit it for you. If you contribute 2/3 more patches you can request commit access yourself.
